### PR TITLE
Used mobility derivative in CHBulk and CHInterface 

### PR DESCRIPTION
**Pull request by tonkmr**
I have now added the mobility derivatives to CHBulk and CHInterface. This work addresses #5167

### CODE REVIEW

**Comment by tonkmr** [comment]
@dschwen can you fix this conflict? I won't have time to get to it until I get back.

**Comment by dschwen** [comment]
Yes, I'll make a pull request against your feature branch.

**Comment by tonkmr** [comment]
Thanks @dschwen!

**Comment by tonkmr** [comment]
The app test failed because this changes requires a patch for MARMOT. The MARMOT patch has been pushed with all but the new hash for the MOOSE submodule. So, since these tests will never pass without the MARMOT patch, but I can't make a merge request for MARMOT until these changes have been merged and I update the hash, what do I do?

**Comment by dschwen** [comment]
MOOSE change gets merged. Marmot submodule won't get automatically updated.
We patch Marmot and manually update the MOOSE submodule in the same PR.
Marmot PR will pass and get merged. Everything is peachy.

**Comment by tonkmr** [comment]
So someone needs to merge this (after comments, changes, etc) with the moose_PR_app_tests failing?

**Comment by permcody** [comment]
Yes, That's the plan. Devel is allowed to break, it doesn't break any
applications.

**Comment by permcody** [comment]
@tonkmr do you have your patches ready? If so, I'll merge

**Comment by tonkmr** [comment]
The patches are ready, but make sure that Daniel and you have reviewed the code before you merge.

File: modules/phase_field/src/kernels/CHParsed.C [hunk 1]
```diff
@@ -54,13 +56,13 @@ CHParsed::computeGradDFDCons(PFFunctionType type)
...
-        res += (*_grad_vars[i])[_qp] * (*_second_derivatives[i])[_qp];
+        res += (*_grad_vars[i])[_qp]*(*_second_derivatives[i])[_qp];
       return res;
 
     case Jacobian:
-      res = _grad_phi[_j][_qp] * (*_second_derivatives[0])[_qp];
+      res = _grad_phi[_j][_qp]*(*_second_derivatives[0])[_qp];
       for (unsigned int i = 0; i <= _nvar; ++i)
-        res += _phi[_j][_qp] * (*_grad_vars[i])[_qp] * (*_third_derivatives[i])[_qp];
+        res += _phi[_j][_qp]*(*_grad_vars[i])[_qp]*(*_third_derivatives[i])[_qp];
```
Comment (permcody) [inline modules/phase_field/src/kernels/CHParsed.C]: There are several places in this PR where you collapse the space around the binary operators. In general we encourage you to leave space around operators. It's part of the code standard and it helps with readability. Other than that I think this PR looks fine.

File: modules/phase_field/include/kernels/CHInterface.h [hunk 2]
```diff
@@ -28,24 +30,33 @@ class CHInterface : public Kernel
...
 
+  ///Variables for second order derivatives
   VariableSecond & _second_u;
   VariableTestSecond & _second_test;
   VariablePhiSecond & _second_phi;
+
+  ///Number of variables
+  unsigned int _nvar;
+
+  ///Material property vectors
```
Comment (dschwen) [inline modules/phase_field/include/kernels/CHInterface.h]: @andrsd would call this _Captain Obvious_... ...me, too.

By the way you can make Doxygen doc comments apply to multiple members by wrapping them with ```@{ @}``` like this:
```
///@{
/// Mobility derivatives
std::vector<const MaterialProperty<Real> *> _dMdarg;
std::vector<const MaterialProperty<Real> *> _d2Mdcdarg;
std::vector<std::vector<const MaterialProperty<Real>* > > _d2Mdargdarg;
///@}
```

File: modules/phase_field/src/kernels/CHInterface.C [hunk 3]
```diff
@@ -13,54 +13,97 @@ InputParameters validParams<CHInterface>()
...
+    //Set coupled variable gradients
+    _coupled_grad_vars[i] = &coupledGradient("args", i);
+  }
 }
 
 Real
 CHInterface::computeQpResidual()
 {
-  return _kappa[_qp] * _second_u[_qp].tr() * (_M[_qp] * _second_test[_i][_qp].tr() + _grad_M[_qp] * _grad_test[_i][_qp]);
+  RealGradient grad_M = _dMdc[_qp]*_grad_u[_qp];
```
Comment (dschwen) [inline modules/phase_field/src/kernels/CHInterface.C]: I really don't like the loss of whitespace around the operators throughout this entire change set.

File: modules/phase_field/src/kernels/CHParsed.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/kernels/CHParsed.C]: I have to touch these files in my big PR anyways, so unless @tonkmr is vehemently opposed to it I could restore them.

File: modules/phase_field/src/kernels/CHInterface.C [hunk 4]
```diff
@@ -13,54 +13,97 @@ InputParameters validParams<CHInterface>()
...
+
+  for (unsigned int i = 0; i < _nvar; ++i)
+    dgrad_Mdarg += (*_d2Mdargdarg[cvar][i])[_qp]*_phi[_j][_qp]*(*_coupled_grad_vars[cvar])[_qp];
+
+  //Jacobian value using product rule
+  Real value = _kappa[_qp]*_second_u[_qp].tr()*((*_dMdarg[cvar])[_qp]*_phi[_j][_qp]*_second_test[_i][_qp].tr()
+                                                + dgrad_Mdarg*_grad_test[_i][_qp]);
+
+  return value;
+
```
Comment (dschwen) [inline modules/phase_field/src/kernels/CHInterface.C]: Empty line, but I can remove that in my PR later

File: modules/phase_field/src/kernels/CHParsed.C (hunk 1 above)
Comment (tonkmr) [inline modules/phase_field/src/kernels/CHParsed.C]: I like to space the math operators in terms of order of operations. So, I have no space around the * or / and I leave a space on either side of the + and -. It makes it much easier for me to quickly parse when I look at a line of math.

File: modules/phase_field/src/kernels/CHInterface.C (hunk 3 above)
Comment (tonkmr) [inline modules/phase_field/src/kernels/CHInterface.C]: If the standard is to always leave a space around the operators, that is OK. But I much prefer it the way that I described above.

File: modules/phase_field/src/kernels/CHParsed.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/kernels/CHParsed.C]: But there are only multiplications in this term for example and if you have
competing access [] and indirection * operators I find it a lot harder to
parse.

**Comment by tonkmr** [comment]
I can make the changes, but it might not be until much later.

**Comment by permcody** [comment]
No need - we merged it as is, it'll be a different PR

**Comment by dschwen** [comment]
I have added the changes in my PR in 760619d00f19. I ended up formatting the terms on multiple lines to groups sums. I'm thinking of building an atom plugin to auto-format math expressions with indentation.

**Pull request merged**